File: Poissonian_data/hidden_poiss_data.m
close all
clear all

y=zeros(1,13);
lambda=1;
for i = 1 : 13
    while(y(i)<4 )
        y(i) = poissrnd(lambda,1,1);
    end
end
niter=2000;
lambdas=zeros(1, niter); % for the  non-Rao-Blackwell estimates
lambdasest=zeros(1, niter); % for the  Rao-Blackwell estimates

for j = 1 :niter
   lambda= gamrnd(313+sum(y), 1/360, 1,1);
   y=zeros(1,13);
   for i = 1 : 13
      while(y(i)<4 )
        y(i) = poissrnd(lambda,1,1);
       end
   end
   lambdahat=(313+sum(y))/360;
   lambdasest(j)=lambdahat;
   lambdas(j)=lambda;
end


% Convergence
lambdabarRB=zeros(1,niter);
lambdabar=zeros(1,niter);
for i = 1:niter 
		lambdabarRB(i) = mean(lambdasest(1:i));
        lambdabar(i) = mean(lambdas(1:i));
end
figure(1);plot(1:niter,lambdabar);xlabel('iterations');ylabel('average'); 
hold on
plot(1:niter,lambdabarRB, 'r')
legend('usual','R-B')
title('Evolution of the R-B empirical average')


% Estimates and histograms
lambdaestRB=mean(lambdasest(1000:2000))  %Rao-Blackwell estimate
lambdaest=mean(lambdas(1000:2000)) % Usual estimate

figure; hist(lambdasest(1000:2000),6)
figure; hist(lambdas(1000:2000),6)

File: Poissonian_data/hidden_poiss_data.R
# Initialization 
y=rep(0,13)
lambda=1
for (i in 1:13)
{while(y[i]<4) y[i]=rpois(1, lambda)}
niter=2000
lambdas=rep(0,niter) #for the Non-Rao-Blackwell estimates
lambdasest=rep(0,niter) #for the Rao-Blackwell estimates
 for(j in 1:niter)
 {
   
 lambda=rgamma(1,shape = 313+sum(y), scale = 1/360) 
 y=rep(0,13)
 for (i in 1:13)
  {while(y[i]<4) y[i]=rpois(1, lambda)}
  lambdahat=(313+sum(y))/360
  lambdasest[j]=lambdahat
  lambdas[j]=lambda
 }
 
 # Convergence 
 
 par(mfrow=c(1,3))
 moyLam=cumsum(lambdasest)/(1:niter) # Empir avg of Rao Blackwell estim
 moylam=cumsum(lambdas)/(1:niter) #Empirical average of lambdas
 plot(moylam, type='l', ylab='', xlab="iterations")
 lines(moyLam, col='red')
 
 # Post burn-in estimates and histograms
 mean(lambdasest[1000:2000]) # Rao-Blackwell estimate
 mean(lambdas[1000:2000]) # Usual estimate
 
 hist(lambdas[1000:2000],main='Empirical Average', xlab="")
 hist(lambdasest[1000:2000],main='Rao-Blackwell', xlab="")
 par(mfrow=c(1,1))
